mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -2457,18 +2457,19 @@ int Send_nochecks (byte [] buf, int offset, int size, SocketFlags flags, out Soc
 			}
 
 			int nativeError;
-			int ret = Send_internal (safe_handle, buf, offset, size, flags, out nativeError);
-
-			error = (SocketError)nativeError;
-
-			if (error != SocketError.Success && error != SocketError.WouldBlock && error != SocketError.InProgress) {
-				is_connected = false;
-				is_bound = false;
-			} else {
-				is_connected = true;
-			}
+			int sent = 0;
+			do {
+				sent += Send_internal (safe_handle, buf, offset + sent, size - sent, flags, out nativeError);
+				error = (SocketError)nativeError;
+				if (error != SocketError.Success && error != SocketError.WouldBlock && error != SocketError.InProgress) {
+					is_connected = false;
+					is_bound = false;
```

Comment:
We need to `break` or `return` here, or in case of error we will keep trying to send more data.